etna/core/fields.py

```diff
@@ -0,0 +1,236 @@
+from datetime import date
+from typing import Optional, Sequence, Union
+
+from django import forms
+from django.core.exceptions import ValidationError
+from django.utils.translation import gettext_lazy as _
+
+from .validators import PositiveIntegerStringValidator
+from .widgets import DateInputWidget
+
+END_OF_MONTH = "END_OF_MONTH"
+
+
+class DateInputField(forms.MultiValueField):
+    """
+    .. _Date input: https://design-system.service.gov.uk/components/date-input/
+
+    Copied from crispy-forms-gds, and extended to support default value substitution
+    for individual date segments when only a partial date is entered. We are also
+    using a more specific validator for each sub-field to catch any obvious issues
+    before attempting to combine the values.
+
+    The original definition can be found at:
+    https://github.com/wildfish/crispy-forms-gds/blob/master/src/crispy_forms_gds/fields.py
+    """
+
+    widget = DateInputWidget
+
+    def __init__(
+        self,
+        default_day: Optional[Union[int, str]] = None,
+        default_month: Optional[int] = None,
+        default_year: Optional[int] = None,
+        **kwargs,
+    ):
+        # Default field values are stored in a 3-item tuple, in an order that
+        # corresponds with the field's sub-fields: day, month and year.
+        # Values are here may be int or None values, or in the case of `default_day`,
+        # an "END_OF_MONTH" could be provided, which substitues a missing day value
+        # for the end of the entered/substituted month value.
+        self.default_values = (default_day, default_month, default_year)
+        if default_day or default_month or default_year:
+            kwargs["require_all_fields"] = False
+        else:
+            kwargs["require_all_fields"] = True
+
+        fields = (
+            forms.CharField(
+                label=_("Day"),
+                error_messages={"incomplete": "Enter the day of the month"},
+                validators=[
+                    PositiveIntegerStringValidator(
+                        min=1,
+                        max=31,
+                        msg="Entered date must be a real date, for example 23 9 2017.",
+                    )
+                ],
+                required=default_day is None,
+            ),
+            forms.CharField(
+                label=_("Month"),
+                error_messages={"incomplete": "Entered date must include a Month."},
+                validators=[
+                    PositiveIntegerStringValidator(
+                        min=1,
+                        max=12,
+                        msg="Entered date must be a real date, for example 23 9 2017.",
```

Comment:
These are component-specific errors, so making them all generic makes the field less helpful. What was wrong with the original wording?

Reply:
There's definitely no reason the 'Day', 'Month' or 'Year' to be capitalised in any of these messages

Reply:
This the requirement.